crates/kubelet/src/device_plugin_manager/mod.rs

```diff
@@ -0,0 +1,954 @@
+                        current_devices.iter().for_each(|(_, device)| {
+                            // (1) Device modified or already registered
+                            if let Some(previous_device) = previous_endpoint_devices.get(&device.id)
+                            {
+                                if previous_device.health != device.health {
+                                    all_devices
+                                        .lock()
+                                        .unwrap()
+                                        .get_mut(&list_and_watch_resource_name)
+                                        .unwrap()
+                                        .insert(device.id.clone(), device.clone());
+                                    update_node_status = true;
+                                } else if previous_device.topology != device.topology {
+                                    // TODO: how to handle this
+                                    error!("device topology changed");
+                                }
+                            // (2) Device added
+                            } else {
+                                let mut all_devices_map = all_devices.lock().unwrap();
+                                match all_devices_map.get_mut(&list_and_watch_resource_name) {
+                                    Some(resource_devices_map) => {
+                                        resource_devices_map
+                                            .insert(device.id.clone(), device.clone());
+                                    }
+                                    None => {
+                                        let mut resource_devices_map = HashMap::new();
+                                        resource_devices_map
+                                            .insert(device.id.clone(), device.clone());
+                                        all_devices_map.insert(
+                                            list_and_watch_resource_name.clone(),
+                                            resource_devices_map,
+                                        );
+                                    }
+                                }
+                                update_node_status = true;
+                            }
+                        });
+
+                        // (3) Check if Device removed
+                        previous_endpoint_devices
+                            .iter()
+                            .for_each(|(_, previous_device)| {
+                                if !response.devices.contains(previous_device) {
+                                    // TODO: how to handle already allocated devices? Pretty sure K8s lets them keep running but what about the allocated_device map?
+                                    all_devices
+                                        .lock()
+                                        .unwrap()
```

Comment:
I have been a little confused about when to use tokio's Mutex. I'd love if you could clarify the use scenarios. Tokio's documentation [states](https://docs.rs/tokio/0.2.25/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use) that:
"If the value behind the mutex is just data, it’s usually appropriate to use a blocking mutex such as the one in the standard library or parking_lot." All of the device manager's fields that are wrapped in a mutex are data rather than fitting into the "primary use case for the async mutex is to provide shared mutable access to IO resources such as a database connection." Since "the async mutex more expensive than the blocking mutex" is there a way I can restructure access to the data such that std mutex can be used? Or is the idea to use the `tokio::std::Mutex` it if the data is ever accessed in an async function or closure that is awaited upon? Is that what "across await points" means? [This example](https://github.com/tokio-rs/mini-redis/blob/master/src/db.rs#L111) that tokio points to uses the `std::sync::Mutex` inside an async function that is called without `await`ing it.

Reply:
Resolving this as we talked offline about it